review-a-pr: improve content (#37)
zz-jason/tidb-dev-guide: contribute-to-tidb/review-a-pr-stage.md

# How to Review a Pull Request

> This guide is for all contributors that want to help with reviewing code contributions. Thank you for your effort - good reviews are one of the most important and crucial parts of an open source project. This guide should help the community to make reviews such that:

Author: Updated.

* Contributors have a good contribution experience.
* Our reviews check all important aspects of a contribution.

> * We make sure to keep a high code quality in TiDB.

Author: Done.

* We avoid situations where contributors and reviewers spend a lot of time refining a contribution that gets rejected later.

> ## Things to know before the code review

Author: Thanks for your suggestion. What do you mean by "don't need this section"? Do you mean that we remove the whole section or keep the content but avoid as a separated section?

> ## Things to know before the code review

Author: Waiting response. I'll continue the merging phase. If you have further concern, please open an issue.

Why is code review important?

* Eliminate bugs early
* Make the project maintainable in the long run

Some high level guidelines about code review in the TiDB community:

### Code review is more important than writing your own PR

> Modern software development is about cooperation, and nowadays software is no longer one single man’s piece of art. To speed up the development and make the processing scale, it is important to avoid blocking. If the members in an organization give higher priority for writing their own code, others waiting for code review get blocked, and there will be a lot of pending PRs for review. Eventually most of the developers get blocked, the whole team slows down.

Author: Updated.

> ### It's the author's responsibility to get the PR merged

Author: Resolved by https://github.com/zz-jason/tidb-dev-guide/pull/37/files#r655986145

> We had a fierce debate on how to speed up the review process long long ago. What the role the reviewer and PR owner should play and what is each of their responsibility to get the PR merged faster.

Author: Resolved by https://github.com/zz-jason/tidb-dev-guide/pull/37/files#r655986145

> After the meeting, most of the participants agree with that the PR owner is responsible for getting PR merged.

Author: Resolved by https://github.com/zz-jason/tidb-dev-guide/pull/37/files#r655986145

If I am the PR author, I am the one who wants to merge the PR into the repository most. I'm willing to lead that change. So I'm supposed to do my best to have someone review the PR. The reviewers should review the PR in time and give feedback, but it is not their responsibility to push the PR to be merged.

### 2 Approvals for a PR to be accepted

> For most of the repositories in the TiDB community, we require 2 approvals before the PR is accepted. A few repositories require different number of approvals, but 2 approvals is the default setting. 2 approvals ensures that even after the original PR author's resignation, there is someone familiar with the code and the project still keeps moving. And if the author really resigns, the code reviewer is the first person responsible for the code.

Author: Updated.

> ### Never give approval to to the code you don't understand

Author: ```suggestion
### Never approve code you don't understand
```

> ### Never give approval to to the code you don't understand

Author: Resolved by https://github.com/zz-jason/tidb-dev-guide/pull/37/files#r655991259

> A review should be thorough enough that the reviewer could explain the change at a reasonable level of detail to another developer. This ensures that the details of the code base are known to more than a single person. Remember: you are responsible for the code you reviewed, it is all about your reputation.

Author: Updated.

> A review should be thorough enough that the reviewer could explain the change at a reasonable level of detail to another developer. This ensures that the details of the code base are known to more than a single person. Remember: you are responsible for the code you reviewed, it is all about your reputation.

Author: Updated.

> ### No more than 300 lines for each PR

Author: ```suggestion
### PRs must be 300 lines or less
```

> We encourage quick, small changes. 300 lines is basically normal people can comprehend within an hour. The reviewer has the right to ask the PR author to split the PR if it is too large. For the PR that is hard to split, the code reviewers can ask the author to have a meeting in person to explain the changes, until they fully understand the changes.

Author: @tomdewan At least it is the best practice so far with `pingcap/tidb` codebase.

> We encourage quick, small changes. 300 lines is basically normal people can comprehend within an hour. The reviewer has the right to ask the PR author to split the PR if it is too large. For the PR that is hard to split, the code reviewers can ask the author to have a meeting in person to explain the changes, until they fully understand the changes.

Author: ... if it is 300 lines of logic code. But yes, it depends.

@zz-jason @tiancaiamao is 300 lines baseline too tough and subjective?

> We encourage quick, small changes. 300 lines is basically normal people can comprehend within an hour. The reviewer has the right to ask the PR author to split the PR if it is too large. For the PR that is hard to split, the code reviewers can ask the author to have a meeting in person to explain the changes, until they fully understand the changes.

Author: https://github.com/zz-jason/tidb-dev-guide/pull/37/files#r656069172

With an offline discussion with @zz-jason we agree with that 500 lines is the current guideline. And as it is a guideline, not a requirement, I also update the words.

### One PR just does one thing

> To make the PR easier to understand and get merged faster, this is also highly encouraged.

Author: ```suggestion
To make the PR easier to understand and get merged faster, it should focus on one thing.
```

> For example, a bugfix should be in a separated PR, a refactoring should be in a separate PR. No matter how small it is, like updating the package version, or even one line changes to fix a typo, the change does exactly one thing and gets it right.

Author: Updated.

### Principles of the code review

* Technical facts and data overrules opinions and personal preferences

> * Software design is about trade-off and there is no silver bullet

Author: ```suggestion
* Software design is about trade-offs, and there is no silver bullet.
```

Everyone comes from different technical backgrounds with different knowledge. They have their own personal preferences. It is important that the code review is not based on biased opinions.

> A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.

Author: Updated.

> A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.

Author: Updated.

> A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.

Author: ```suggestion
A common mistake is to focus too much on style. Style is important: that's why there are style guides. Just let the bot do most of the trivial work. Each PR reviewer is a human who is supposed to focus on things beyond what a bot can do. When you are reviewing a PR, make sure that the changes do not break the program logic or increase the code base maintenance cost.
```

> A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.

Author: Right. The adjective is inverted.

> A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.

Author: Updated.

Sometimes, making choices can be tricky.

> Said that there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, which means a huge maintenance burden. Shall we accept or reject it?

Author: ```suggestion
For example, let's say there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path, and every subsequent feature must consider it. This creates a huge maintenance burden. Shall we accept or reject it?
```

> Said that there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, which means a huge maintenance burden. Shall we accept or reject it?

Author: Yep.

> Said that there is a critical bug and a PR is read to fix it, but the change is risky to introduce other kind of bugs. Shall we accept or reject it?

Author: ```suggestion
Or let's say there's a critical bug and a PR is read to fix it, but there's a risk the PR will introduce other bugs. Shall we accept or reject it?
```

> Said that there is a critical bug and a PR is read to fix it, but the change is risky to introduce other kind of bugs. Shall we accept or reject it?

Author: Updated.

> The answer is always “it depends”. Software design is more like a kind of art than technology. It is about aesthetics and your taste of the code. Trade-offs are always there and there is no perfect solution sometimes. Still, there is a general standard (from [@shenli](https://github.com/shenli), maybe we can call it "better than before?"):

Author: Updated.

* It is making something better, although not a perfect solution.
* It can mitigate the pain of the user.

> * There is no bug introduced by the change itself.

Author: Updated.

> Then the code can be accepted.

Author: ```suggestion
Then, we can accept the code.
```

> Then the code can be accepted.

Author: Updated.

Of course, code review should have some minimal requirements.

> For a bug fix PR, if it comes with no test case, the reviewer should ask the author to add tests.

Author: ```suggestion
If a bug fix PR doesn't include a test case, the review should ask the author to add tests.
```

> For a performance improvement PR, if no benchmark result is provided, the reviewer should ask the author to attach a benchmark result.

Author: ```suggestion
If a performance improvement PR doesn't include a benchmark result, the reviewer should ask the author to attach one.
```

> For a performance improvement PR, if no benchmark result is provided, the reviewer should ask the author to attach a benchmark result.

Author: Updated.

## What to look for in a code review

### Design

The most important thing to cover in a review is the overall design.

Read the description of the PR, and if it's not easy to understand, ask the author to improve it. You may consider questions like:

* Does the change integrate well with the rest of our system?

> * Is it a good time now to add this functionality?

Author: ```suggestion
* Is now a good time to add this functionality?
```

[TiDB Internals forum](https://internals.tidb.io/) is a place for discussion. For pull requests that are created without prior consensus, try to seek consensus there.

### Functionality

> * Does this PR do what the developer intend to do?

Reviewer: * Does this PR do what the developer intends it to do?

Author: ```suggestion
* Does this PR do what the developer intends it to do?
```

> * Is what the developer intend to do good for the users with this implementation?

Reviewer: * How will implementing this PR help the user?

Author: ```suggestion
* How will implementing this PR help the user?
```

> * Is what the developer intend to do good for the users with this implementation?

Author: Updated.

> Basically, PR authors are expected to test their PRs well-enough before code reviews. However, as a reviewer you should still think about edge cases, especially for problems about concurrency.

Reviewer: ... to sufficiently test their PRs before code reviews. However, as a reviewer, you should still think about edge cases -- especially for problems about concurrency.

Author: ```suggestion
Basically, PR authors are supposed to sufficiently test their PRs before code reviews. However, as a reviewer, you should still think about edge cases, especially for problems about concurrency.
```

> Basically, PR authors are expected to test their PRs well-enough before code reviews. However, as a reviewer you should still think about edge cases, especially for problems about concurrency.

Author: Updated.

You can always validate the PR if you want. It's hard to understand how some changes will impact a user when you're just reading the code, such as a UI change. You can ask the PR author to give you a demo of the functionality if it's too inconvenient to try it yourself.

### Logic

> In the case that concurrent programming involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.

Reviewer: When concurrent programming is involved, ...

Author: ```suggestion
When concurrent programming is involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.
```

> The human brain is just like a special machine, it is good at checking a small piece of logic before the states become too complex. Complexity is the biggest challenge of software.

Reviewer: I think you can safely delete this paragraph. Remember, your reader's time is limited.

Author: Updated.

> The human brain is just like a special machine, it is good at checking a small piece of logic before the states become too complex. Complexity is the biggest challenge of software.

Author: Updated.

### Complexity

Is the PR more complex than it should be? Check the following items for review the complexity of a PR:

* Are individual lines too complex?
* Are functions too complex?
* Are classes too complex?

> "Too complex" usually means "can't be understood quickly by readers". It may also mean "developers are likely to introduce bugs when they try to call or modify this code".

Reviewer: ... readers."

... code."

Author: Updated.

> A particular type of complexity is over-engineering, where developers have made the code more generic than it needs to be, or added functionality that isn’t presently needed by the system. Reviewers should be especially vigilant about over-engineering. Encourage developers to solve problems must to be solved now, not problems that a developer speculates might need to be solved in the future. 

Reviewer: ... to solve problems that must be solved now -- not problems ...

Author: ```suggestion
A particular type of complexity is over-engineering, which means that PR authors have made the code more generic than it needs to be, or added functionality that is not presently needed by the system. Reviewers should be especially vigilant about over-engineering. The TiDB community encourages PR authors to solve the problems that must be solved now -- not the problems that might need to be solved in the future. 
```

> A particular type of complexity is over-engineering, where developers have made the code more generic than it needs to be, or added functionality that isn’t presently needed by the system. Reviewers should be especially vigilant about over-engineering. Encourage developers to solve problems must to be solved now, not problems that a developer speculates might need to be solved in the future. 

Author: Updated.

### Tests

A PR should be test covered, whether the tests are unit tests, integration tests, or end-to-end tests.

Ensure the tests in the PR are correct, sensible, and useful.

* Are they covering interesting cases?
* Are they readable?
* Does the PR lower overall test coverage? 

Simulate cases where the code may break.

* Will the tests actually fail when the code is broken?

> * If the code changes beneath them, will they start producing false positives?

Author: They mean the current change set. Let me try to rephrase this sentence.

> * If the code changes beneath them, will they start producing false positives?

Author: Actually it is almost the same meaning as

```
* Will the tests actually fail when the code is broken?
```

* Does each test makes simple and useful assertions?
* Are the tests separated appropriately between different test methods?

> Remember that tests are also code that have to be maintained, and thus do not accept unnecessary complexity in tests.

Reviewer: Comment: In some cases, there are two spaces between words. Only use one.

Author: Ok. I will do a global search and replace.

> Last but not least, keep an eye on how long a single unit takes to run. That is, whether or not the CI is slowed by the tests introduced.

Reviewer: That is, determine whether the test slow down the CI.

### Naming

Does the developer pick good names for everything? A good name is long enough to fully communicate what the item is or does, without being so long that it becomes hard to read.

### Comments

* Does the developer write clear comments in understandable English?
* Are all of the comments actually necessary?

> Usually comments are useful when they explain why the code exists, while not explain what the code is doing. If the code isn't clear enough to explain itself, then the code should be made simpler.

Reviewer: Usually, ...

Author: Updated.

> Usually comments are useful when they explain why the code exists, while not explain what the code is doing. If the code isn't clear enough to explain itself, then the code should be made simpler.

Reviewer: ... then the code should be simplified.

Author: ```suggestion
Usually, comments are useful when they explain why the code exists, while not explain what the code is doing. If the code isn't clear enough to explain itself, then the code should be simplified.
```

> There are some exceptions; for example, regular expressions and complex algorithms often benefit greatly from comments that explain what they’re doing. But mostly comments are for information that the code itself can hardly contain, like the reason behind a decision.

Reviewer: But comments are mostly for information ...

Author: ```suggestion
There are some exceptions; for example, regular expressions and complex algorithms often benefit greatly from comments that explain what they’re doing. But comments are mostly for information, like the reason behind a decision.
```

It can also be helpful to look at comments existing before this PR. Maybe there is a `TODO` can be removed now, a comment advising against this change being made, etc.

Note that comments are different from documentation of classes, modules, or functions, which should instead express the purpose of a piece of code, how it should be used, and how it behaves when used.

### Style

> Make sure the PR follows our [code style and quality guide](code-style-and-quality-guide.md). For Go and Rust, there are built in tools with the compiler toolchain.

Reviewer: ... built-in ...

If you want to improve style points outside the style guide, you can comment that you think it would improve the code readability but isn't mandatory. Don't block PRs from being submitted based only on personal style preferences.

> The PR author should not include major style changes combined with other changes. For example, if a PR wants to reformat the whole file should not make functional changes. This has been explained as "One PR just does one thing" in the previous section.

Reviewer: The PR author should combine major style changes -- such as reformatting the entire file -- with functional changes. This would violate the guideline covered earlier: one PR does one thing.

Author: ```suggestion
The PR author should not combine nontrivial style changes with functional changes. This would violate the guideline covered earlier: one PR does one thing.
```

### Consistency

What if the existing code is inconsistent with the style guide? The style guide is the absolute authority: if something is required by the style guide, the PR should follow the guidelines.

> In some cases, the style guide makes recommendations rather than declaring requirements. In these cases, it’s a judgment call whether the new code should be consistent with the recommendations or the surrounding code. Bias towards following the style guide unless the local inconsistency would be too confusing

Reviewer: Consider changing the last sentence to, "When in doubt, lean towards following the style guide -- unless the "local" inconsistency would be too confusing."

Author: Updated.

If no other rule applies, the author should maintain consistency with the existing code.

Either way, encourage the author to file a bug and add a TODO for cleaning up existing code.

### Documentation

> If a PR changes how users build, test, interact with, or release code, the reviewer are supposed to check whether it also updates related documentation, such as READMEs, and any generated reference docs. 

Reviewer: ... must check whether it also updates ...

Author: Updated.

If a PR deletes or deprecates code, it should consider whether the documentation should also be deleted.

If documentation is missing, the reviewer should ask for it.

#### Error Handling

Be careful with error or exception handling. This is one of the most common places for bugs.

* Does the test cover the error code path?
* Does a panic make the whole process exit for server services?
* Does the code handle the resource releasing properly?

> * Especially, for code including concurrent goroutines, is there a goroutine leak or memory leak after the error?

Reviewer: Delete the comma after "Especially."

Author: Updated.

> Sometimes it is difficult for a real program to run into the error code path, and even more difficult to reproduce and debug if it really happens. Then it is necessary to mock the error code path. Tools like [failpoint](https://github.com/pingcap/failpoint) are available for that purpose.

Reviewer: ... Then, ...

### Performance

> Does the PR impose any performance impact? Sometimes a minor change on the hot path that seems unrelated may still affect performance. Is the algorithm in the PR suitable for that case?

Reviewer: Does the PR impact performance? Sometimes, a seemingly minor, unrelated change on the hot may still affect performance.

Author: ```suggestion
Does the PR impact performance? Sometimes, a seemingly minor, unrelated change on the hot may still affect performance.

Is the algorithm in the PR suitable for that case?
```

> Does the PR impose any performance impact? Sometimes a minor change on the hot path that seems unrelated may still affect performance. Is the algorithm in the PR suitable for that case?

Author: Updated.

### Offer sincere praise

Most reviewers focus only on what’s wrong with the code, but they should offer encouragement and appreciation for good practices, as well. It’s sometimes even more valuable, in terms of mentoring, to tell a developer what they did right than to tell them what they did wrong.

For example, imagine you’re reviewing for an author who struggles to write documentation, and you come across a clear, concise function comment. Let them know they nailed it. They will improve faster if you tell them when they got it right instead of just waiting to ding them when they screw up.

#### Summary

* Be kind to the coder, not to the code.
* Ask questions rather than make statements.
* Treat people who know less than you with respect, deference, and patience.
* Remember to praise when the code quality exceeds your expectation.
* It isn't necessarily wrong if the coder's solution is different from yours.
* Refer to the code style document when necessary.

## Review with the @ti-chi-bot

> The TiDB community is using a service called [@ti-chi-bot](https://book.prow.tidb.io/#/en/) to help with the review of the pull requests.

Reviewer: ... community uses ...

Author: ```suggestion
The TiDB community uses a service called [@ti-chi-bot](https://book.prow.tidb.io/#/en/) to help with the review of the pull requests.
```

> You can see the full list of commands accepted by the bot in [this page](https://prow.tidb.io/command-help?repo=pingcap%2Ftidb).

Reviewer: ... by the bot at ...

Author: ```suggestion
You can see the full list of commands accepted by the bot at [this page](https://prow.tidb.io/command-help?repo=pingcap%2Ftidb).
```

> Here are the most frequent commands you are like gonna use during a review.

Reviewer: Did you mean "you're gonna like" or "you're gonna use"?

... during a review:

Author: ```suggestion
Here are the most frequent commands you're gonna use during a review:
```

## Things to know before the code review

Why is code review so important?

* A good review can eliminate bugs early.
* A good review can make the project maintainable in the long run.

To conduct a good review, it is always helpful to read the following high-level guidelines about code review in the TiDB community:

### Code review is more important than writing your own PR

Nowadays software development is no longer a single person’s piece of art but the result of team cooperation. To speed up the development and make the processing scale, it is important to avoid blocking. If all developers in an organization prioritize writing code over reviewing code, more and more PRs waiting for code review get blocked, and the whole team slows down.

### It's the PR author's responsibility to get the PR merged

Because PR authors are the ones who want to merge the PRs into the repository the most, after an offline discussion we agree on the following:

* PR authors are responsible for getting PRs reviewed and merged.
* PR reviewers are responsible for reviewing PRs and providing feedback in time but not responsible for pushing the PR to be merged.

> ### 2 Approvals for a PR to be accepted

Author: ```suggestion
### Two Approvals for a PR to be accepted
```

* The review keeps the high quality of TiDB code.
* Reviewers cover all important aspects of a contribution
* Contributors have a good contribution experience..
* The review avoids situations in which contributors and reviewers spend a lot of time refining a contribution that gets rejected later.

## Things to know before the code review

Why is code review so important?

* A good review can eliminate bugs early.
* A good review can make the project maintainable in the long run.

To conduct a good review, it is always helpful to read the following high-level guidelines about code review in the TiDB community:

### Code review is more important than writing your own PR

Nowadays software development is no longer a single person’s piece of art but the result of team cooperation. To speed up the development and make the processing scale, it is important to avoid blocking. If all developers in an organization prioritize writing code over reviewing code, more and more PRs waiting for code review get blocked, and the whole team slows down.

### It's the PR author's responsibility to get the PR merged

> We had a fierce debate on how to speed up the review process long long ago. What the role the reviewer and PR owner should play and what is each of their responsibility to get the PR merged faster.
> 
> After the meeting, most of the participants agree with that the PR owner is responsible for getting PR merged.

Author: ```suggestion
Because PR authors are the ones who want to merge the PRs into the repository the most, after an offline discussion we agree on the following:
```

### It's the PR author's responsibility to get the PR merged

Because PR authors are the ones who want to merge the PRs into the repository the most, after an offline discussion we agree on the following:

* PR authors are responsible for getting PRs reviewed and merged.
* PR reviewers are responsible for reviewing PRs and providing feedback in time but not responsible for pushing the PR to be merged.

### Two Approvals for a PR to be accepted

In the TiDB community, most repositories require two approvals before a PR can be accepted. A few repositories require a different number of approvals, but two approvals are the default setting. 

Two approvals ensure that even the original PR author resigns, there is still someone who is familiar with the code and can keep the project moving. And if the author really resigns, the code reviewer becomes the first person who is responsible for the code.

### Never approve code you don't understand

> A review should be thorough enough that the reviewer could explain the change at a reasonable level of detail to another developer. This ensures that the details of the code base are known to more than a single person. Remember: you are responsible for the code you reviewed, it is all about your reputation.

Author: ```suggestion
A review should be thorough enough. After the review, you should be able to explain the change details at a reasonable level to other developers. This ensures that the details of the codebase are known to more than a single person. Remember: you are responsible for the code you review; it is all about your reputation.
```

> A review should be thorough enough. After the review, you should be able to explain the change details at a reasonable level to other developers. This ensures that the details of the codebase are known to more than a single person. Remember that you are responsible for the code you reviewed because it is all about your reputation.

Author: ```suggestion
A review should be thorough enough. After the review, you should be able to explain the change details at a reasonable level to other developers. This ensures that the details of the codebase are known to more than a single person. Remember: you are responsible for the code you review; it is all about your reputation.
```

* PR authors are responsible for getting PRs reviewed and merged.
* PR reviewers are responsible for reviewing PRs and providing feedback in time but not responsible for pushing the PR to be merged.

### Two Approvals for a PR to be accepted

In the TiDB community, most repositories require two approvals before a PR can be accepted. A few repositories require a different number of approvals, but two approvals are the default setting. 

Two approvals ensure that even the original PR author resigns, there is still someone who is familiar with the code and can keep the project moving. And if the author really resigns, the code reviewer becomes the first person who is responsible for the code.

### Never approve code you don't understand

A review should be thorough enough. After the review, you should be able to explain the change details at a reasonable level to other developers. This ensures that the details of the codebase are known to more than a single person. Remember: you are responsible for the code you review; it is all about your reputation.

> ### PRs must be 300 lines or less
> 
> The TiDB community encourages quick and small changes in each PR. Three hundred lines is basically what normal people can comprehend in an hour.
> 
> If a PR is too large, you can request the PR author to split the PR. If a PR is too large but hard to split, you can request the author to have a meeting in person to explain the changes, until you fully understand the changes.

Author: ```suggestion
### PRs are suggested to be less than 500 lines

The TiDB community encourages small changes in each PR and sets up five hundred lines as a guideline of PR  size.

If a PR is too large, you can request the PR author to split the PR. If a PR is too large but hard to split, you can request the author to have a meeting in person to explain the changes, until you fully understand the changes.
```

Sometimes, making choices can be tricky.

Said that there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, which means a huge maintenance burden. Shall we accept or reject it?

Said that there is a critical bug and a PR is read to fix it, but the change is risky to introduce other kind of bugs. Shall we accept or reject it?

The answer is always “it depends”. Software design is more like a kind of art than technology. It is about aesthetics and your taste of the code. Trade-offs are always there and there is no perfect solution sometimes. Still, there is a general standard (from [@shenli](https://github.com/shenli), maybe we can call it "better than before?"):

* It is making something better, although not a perfect solution.
* It can mitigate the pain of the user.
* There is no bug introduced by the change itself.

Then the code can be accepted.

> Of course, code review should have some minimal requirements.

Author: ```suggestion
Besides the "better than before?" standard, you should also check some minimal requirements of code reviews.
```

To learn the overall design of a PR, you can read the description of the PR. If the description is unclear or hard to understand, you can request the author to improve it. You may consider asking questions such as the following:

* Does the change integrate well with the rest of our system?
* Is now a good time to add this functionality?

If you need a further discussion of the PR, you can go to [TiDB Internals forum](https://internals.tidb.io/). For pull requests that are created without prior consensus, try to seek consensus there too.

### Functionality

* Does this PR implement what the PR author intends to do?
* How will implementing this PR help the user?

Basically, PR authors are supposed to sufficiently test their PRs before code reviews. However, as a reviewer, you should still think about edge cases, especially for problems about concurrency.

You can always validate the PR if you want. 

For some changes such as UI, it is hard to understand how the changes will impact users when you are just reading the code. In this case, if it is too inconvenient to try it yourself, you can ask the PR author to give you a demo of the functionality.

### Logic

> When concurrent programming is involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.

Author: ```suggestion
When concurrent programming is involved, it is particularly important to review the logic carefully, because the concurrent issues are very hard to detect by just running the code.
```

### Tests

A PR should be test covered, whether the tests are unit tests, integration tests, or end-to-end tests.

To ensure the tests in the PR are correct, sensible, and useful, you can check the following:

* Whether the tests cover all interesting cases?
* Are the test cases readable?
* Does the PR lower the overall test coverage? 

To ensure the tests in the PR are sufficient, you can simulate cases where the code may break and check the following:

* Will the tests fail when the code is broken?
* Does each test make simple and useful assertions?
* Are the tests separated appropriately between different test methods?

Remember that tests are also code that have to be maintained, and thus do not accept unnecessary complexity in tests.

Last but not least, keep an eye on how long a single unit takes to run. That is, determine whether the test slow down the CI.

### Naming

> Does the developer pick good names for everything? A good name is long enough to fully communicate what the item is or does, without being so long that it becomes hard to read.

Reviewer 2: ```suggestion
Does the PR author pick good names for everything? A good name should be clear enough to fully communicate what the item is or does.
```

Reviewer 2: Is it necessarily being long enough?  Are there any existing naming rules, could we provide a link here to those rules, or just provide an example about what is a good name and what is not?

Author: It is a good point. You can create an issue that we add examples on naming and we can address it later.

### Comments

> * Does the developer write clear comments in understandable English?

Reviewer 2: ```suggestion
* Does the PR author write clear comments in understandable English?
```

* Are all of the comments actually necessary?

> Usually, comments are useful when they explain why the code exists, while not explain what the code is doing. If the code isn't clear enough to explain itself, then the code should be simplified.

Reviewer 2: ```suggestion
Usually, comments are useful to explain why the code exists, while not useful to explain what the code is doing. If the code is not clear enough to explain itself, the PR author should simplify the code instead of adding unnecessary comments.
```

> There are some exceptions; for example, regular expressions and complex algorithms often benefit greatly from comments that explain what they’re doing. But comments are mostly for information, like the reason behind a decision.

Reviewer 2: ```suggestion
However, there are some exceptions. For example, the comments to explain what regular expressions and complex algorithms are doing are greatly helpful. But in other cases, comments are mostly for information, such as a reason behind a decision.
```

> It can also be helpful to look at comments existing before this PR. Maybe there is a `TODO` can be removed now, a comment advising against this change being made, etc.

Reviewer 2: ```suggestion
It can also be helpful to look at the existing comments before this PR. Maybe a `TODO` in the existing comments can be removed now, or there is a comment advising against the current changes, etc.
```

> Note that comments are different from documentation of classes, modules, or functions, which should instead express the purpose of a piece of code, how it should be used, and how it behaves when used.

Reviewer 2: ```suggestion
Note that comments are different from documentation of classes, modules, or functions, which covers the purpose of a piece of code, how the code should be used, and how the code behaves when being used.
```

### Style

Make sure the PR follows our [code style and quality guide](code-style-and-quality-guide.md). For Go and Rust, there are built-in tools with the compiler toolchain.

> If you want to improve style points outside the style guide, you can comment that you think it would improve the code readability but isn't mandatory. Don't block PRs from being submitted based only on personal style preferences.

Reviewer 2: ```suggestion
If you want to improve styles not covered in the style guide, you can comment that you think it would improve the code readability but it is not mandatory. Do not block PRs from being merged based only on personal style preferences.
```

> The PR author should not combine nontrivial style changes with functional changes. This would violate the guideline covered earlier: one PR does one thing.

Reviewer 2: ```suggestion
The PR author should not combine nontrivial style changes and functional changes in one PR, because it violates the guideline covered earlier "one PR does one thing".
```

For example, suppose that you are reviewing a PR created by an author who struggles to write documentation, and you come across a clear, concise function comment in the PR. Then, you are recommended to tell such PR authors that they nailed it. To help the PR authors improve faster, tell the authors when they get it right instead of just waiting to ding them when they screw up.

### Summary

* Be kind to PR authors, not to the code.
* Ask questions instead of making statements.
* Treat people who know code less than you with respect, deference, and patience.
* Offer your praise when the code quality exceeds your expectation.
* It is not necessarily wrong if the solution of a PR author is different from yours.
* Refer to the code style document if necessary.

## Review with the @ti-chi-bot

The TiDB community uses a service called [@ti-chi-bot](https://book.prow.tidb.io/#/en/) to help with the review of the pull requests.

You can see the full list of commands accepted by the bot at [this page](https://prow.tidb.io/command-help?repo=pingcap%2Ftidb).

> Here are the most frequent commands you're gonna use during a review:

Reviewer 2: ```suggestion
Here are the most frequently used commands you're going to use during a review:
```

* `/cc @reviewer` requests a review from the reviewer.
* `/assign @committer` assigns a committer to help merge the pull request

> * GitHub reviewing `Approve` or `Request Changes` from reviewers or committers creates a LGTM or veto to the pull request.

Reviewer 2: ```suggestion
* /`Approve` from reviewers or committers creates a LGTM for the pull request.
* /`Request Changes` from reviewers or committers creates a veto for the pull request.
```

Author: There is no /`Approve`.

Two approvals ensure that even the original PR author resigns, there is still someone who is familiar with the code and can keep the project moving. And if the author really resigns, the code reviewer becomes the first person who is responsible for the code.

### Never approve code you don't understand

A review should be thorough enough. After the review, you should be able to explain the change details at a reasonable level to other developers. This ensures that the details of the codebase are known to more than a single person. Remember: you are responsible for the code you review; it is all about your reputation.

### PRs must be 300 lines or less

The TiDB community encourages quick and small changes in each PR. Three hundred lines is basically what normal people can comprehend in an hour.

If a PR is too large, you can request the PR author to split the PR. If a PR is too large but hard to split, you can request the author to have a meeting in person to explain the changes, until you fully understand the changes.

### One PR does one thing

> To make the PR easier to understand and get merged faster, it should focus on one thing.
> 
> For example, a bug fix should be in one PR and a refactoring in another. No matter how small it is, like updating the package version, or even one line changes to fix a typo, the change does exactly one thing and gets it right.

Author: ```suggestion
To make a PR easier to understand and get merged faster, it should focus on one thing.

For example, a bug fix should be in one PR and a refactoring in another. No matter how small it is, the change does exactly one thing and gets it right.
```

### Principles of the code review

* Technical facts and data overrule opinions and personal preferences.
* Software design is about trade-offs, and there is no silver bullet.

Everyone comes from different technical backgrounds with different knowledge. They have their own personal preferences. It is important that the code review is not based on biased opinions.

A common mistake is to focus too much on style. Style is important: that's why there are style guides. Just let the bot do most of the trivial work. Each PR reviewer is a human who is supposed to focus on things beyond what a bot can do. When you are reviewing a PR, make sure that the changes do not break the program logic or increase the code base maintenance cost.

> Sometimes, making choices can be tricky.
> 
> For example, let's say there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path, and every subsequent feature must consider it. This creates a huge maintenance burden. Shall we accept or reject it?
> 
> Or let's say there's a critical bug and a PR is read to fix it, but there's a risk the PR will introduce other bugs. Shall we accept or reject it?

Author: ```suggestion
Sometimes, making choices of accepting or rejecting a PR can be tricky as in the following situations: 

* Suppose that a PR contains special optimization that can improve the overall performance by 30%. However, the PR introduces a totally different code path, and every subsequent feature must consider it.
* Suppose that a PR is to fix a critical bug, but the change in the PR is risky to introduce other bugs.
```

* The change is making something better, although not a perfect solution.
* The change can mitigate the pains of users.
* The change does not introduce a bug.

Then, we can accept the code.

Besides, you should also check some minimal requirements of code reviews.

If a bug fix PR doesn't include a test case, the review should ask the author to add tests.

If a performance improvement PR doesn't include a benchmark result, the reviewer should ask the author to attach one.

## What to check in a code review

### Design

The most important thing of a review is checking the overall design.

To learn the overall design of a PR, you can read the description of the PR. If the description is unclear or hard to understand, you can request the author to improve it. You may consider asking questions such as the following:

* Does the change integrate well with the rest of our system?
* Is now a good time to add this functionality?

If you need a further discussion of the PR, you can go to [TiDB Internals forum](https://internals.tidb.io/). For pull requests that are created without prior consensus, try to seek consensus there too.

### Functionality

> * Does this PR do what the developer intends it to do?

Author: ```suggestion
* Does this PR implement what the PR author intends to do?
```

* Is now a good time to add this functionality?

If you need a further discussion of the PR, you can go to [TiDB Internals forum](https://internals.tidb.io/). For pull requests that are created without prior consensus, try to seek consensus there too.

### Functionality

* Does this PR do what the developer intends it to do?
* Is what the developer intend to do good for the users with this implementation?

Basically, PR authors are expected to test their PRs well-enough before code reviews. However, as a reviewer you should still think about edge cases, especially for problems about concurrency.

You can always validate the PR if you want. 

For some changes such as UI, it is hard to understand how the changes will impact users when you are just reading the code. In this case, if it is too inconvenient to try it yourself, you can ask the PR author to give you a demo of the functionality.

### Logic

When concurrent programming is involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.

> 
> The human brain is just like a special machine, it is good at checking a small piece of logic before the states become too complex. Complexity is the biggest challenge of software.

Author: ```suggestion
```

Make sense.
